enarx-keep/src/backend/kvm/vm/syscall.rs

```diff
@@ -0,0 +1,40 @@
+// SPDX-License-Identifier: Apache-2.0
+
+mod fixups;
+
+use fixups::get_fixup_table;
+
+use ::x86_64::VirtAddr;
+use memory::Register;
+
+use std::convert::AsMut;
+
+#[derive(Clone, Copy)]
+pub struct SallyPort {
+    /// The host virtual address where the shared syscall page is
+    /// located.
+    pub addr: VirtAddr,
+    /// The host virtual address that corresponds to the start of
+    /// the guest's address space. This offset is added to guest
+    /// physical addresses to get the host virtual address it
+    /// corresponds to.
+    pub offset: u64,
+}
+
+impl AsMut<sallyport::Block> for SallyPort {
+    fn as_mut(&mut self) -> &mut sallyport::Block {
+        let blk = unsafe { self.addr.as_mut_ptr::<sallyport::Block>().as_mut().unwrap() };
+
+        if let Some(fixups) = get_fixup_table(unsafe { u64::from(blk.msg.req.num) }) {
```

Comment:
What do you think of letting the shim doing the fixup when constructing the request? Of course, it would need the translation information first.

Reply:
@haraldh I don't see any issues with this. The host virtual offset can be communicated via the bootinfo struct.

Reply:
The only possible use case that would complicate this slightly is if the host VMM has to `mremap` the address space. That would cause a coherency issue that the host VMM would have to preempt the guest to notify it of. In that case it's a whole bag of worms because the guest would have to revalidate all of its pages, so I don't know if that's a use case we are interested in.